src/js/providers/utils/quality-labels.ts

```diff
@@ -65,28 +79,28 @@ export function findClosestBandwidth(bandwidths, targetBandwidth) {
     return closest;
 }
 
-export function toKbps(bandwidth) {
+export function toKbps(bandwidth: number): number {
     return Math.floor(bandwidth / 1000);
 }
 
 // Use an empty object as the context and populate it like a hash map
-export function hasRedundantLevels(levels) {
+export function hasRedundantLevels(levels: Partial<Level>[]): boolean {
     if (!Array.isArray(levels)) {
         return false;
     }
     return checkForLevelDuplicates(levels, ['height', 'bitrate', 'bandwidth']);
 }
 
-export function hasRedundantLabels(levels) {
+export function hasRedundantLabels(levels: Partial<Level>[]): boolean {
     if (!Array.isArray(levels)) {
         return false;
     }
     return checkForLevelDuplicates(levels, ['label']);
 }
 
-function checkForLevelDuplicates(levels, dupKeys) {
-    return levels.some(function (level) {
-        let key;
+function checkForLevelDuplicates(levels: Partial<Level>[], dupKeys: string[]): boolean {
+    return levels.some(function (this: GenericObject, level: Partial<Level>): boolean | undefined {
```

Comment:
The type of `this` would be the type of `levels` (`levels.some`). This mutates `levels` which is bad. We should declare a hash/dict at the top and use that instead of `this`:
```ts
function checkForLevelDuplicates(levels: Partial<Level>[], dupKeys: string[]): boolean {
    const duplicates: { [key: string] : boolean ] = {};
...
	    duplicates[key] = true;
      return foundDuplicate;
```

Reply:
I thought the type of `this` would be the {} type that's passed in as the second argument to this `some` function?
Regardless, I prefer the `duplicates` const.

Reply:
You are right! I didn’t notice the `thisArg` following the callback. It’s fine as is.